Sources/SotoSignerV4/SigV4a.swift

```diff
@@ -0,0 +1,248 @@
+        0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+        0xBC, 0xE6, 0xFA, 0xAD, 0xA7, 0x17, 0x9E, 0x84,
+        0xF3, 0xB9, 0xCA, 0xC2, 0xFC, 0x63, 0x25, 0x4F,
+    ]
+
+    package static func makeDerivedKey(bytes: [UInt8]) throws -> DerivedKeyResult {
+        assert(bytes.count == 32)
+
+        let comparisonResult = Self.compareConstantTime(lhs: bytes, rhs: Self.s_n_minus_2)
+        if comparisonResult > 0 {
+            return .nextCounter
+        }
+
+        return try .success(P256.Signing.PrivateKey(rawRepresentation: bytes.addingOne()))
+    }
+}
+
+extension [UInt8] {
+
+    package mutating func addOne() {
+        var carry: UInt32 = 1
+        for index in self.indices.reversed() {
+            var digit = UInt32(self[index])
+            digit += carry
+            carry = (digit >> 8) & 0x01
+            self[index] = UInt8(digit & 0xFF)
+        }
+    }
+
+    package consuming func addingOne() -> [UInt8] {
+        var new = consume self
+        new.addOne()
+        return new
+    }
+}
+
+package struct HexEncoding<Base: Sequence> where Base.Element == UInt8 {
+    var base: Base
+
+    package init(_ base: Base) {
+        self.base = base
+    }
+}
+
+extension HexEncoding: Sequence {
+    package typealias Element = UInt8
+
+    package struct Iterator: IteratorProtocol {
+        package typealias Element = UInt8
+
+        var base: Base.Iterator
+        var _next: UInt8?
+
+        init(base: Base.Iterator) {
+            self.base = base
+            self._next = nil
+        }
+
+        package mutating func next() -> UInt8? {
+            switch self._next {
+            case .none:
+                guard let underlying = self.base.next() else {
+                    return nil
+                }
+                let first = underlying >> 4
+                let second = underlying & 0x0F
+                self._next = second.makeBase16Ascii()
+                return first.makeBase16Ascii()
+
+            case .some(let next):
+                self._next = nil
+                return next
+            }
+        }
+    }
+
+    package func makeIterator() -> Iterator {
+        Iterator(base: self.base.makeIterator())
+    }
+}
+
+extension HexEncoding: Collection where Base: Collection {
+    package struct Index: Comparable {
+        package static func < (lhs: HexEncoding<Base>.Index, rhs: HexEncoding<Base>.Index) -> Bool {
+            if lhs.base < rhs.base {
+                return true
+            } else if lhs.base > rhs.base {
+                return false
+            } else if lhs.first && !rhs.first {
+                return true
+            } else {
+                return false
+            }
+        }
+
+        var base: Base.Index
+        var first: Bool
+    }
+
+    package var startIndex: Index {
+        Index(base: self.base.startIndex, first: true)
+    }
+
+    package var endIndex: Index {
+        Index(base: self.base.endIndex, first: true)
```

Comment:
Should `first` be set here?

Reply:
yes. endIndex is the first index that is out of bounds. given that we start with first = true, this is correct.